SAK-41200 Added section column to gradebook

adrianfish opened the pull request.

adrianfish commented:
@bjones86 Assign me as reviewer on your gb PRS, if you want. We're both in the same area at the moment.

bjones86 commented on gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java:
```diff
@@ -1029,6 +1034,17 @@ public GradeSaveResponse saveExcuse(final Long assignmentId, final String studen
+		for (CourseSection cs : sectionManager.getSections(site.getId())) {
+			for (EnrollmentRecord er : sectionManager.getSectionEnrollments(cs.getUuid())) {
+				String userId = er.getUser().getUserUid();
+				if (!userSections.containsKey(userId)) {
```
I think it would be more efficient to just do the `.get()` and check for `null` return, rather than doing both a `containsKey()` and a `.get()` (you're effectively doing two bucket searches, rather than one). For example:

```
List<String> sections = userSections.get(userId);
if (sections == null)
{
    sections = new ArrayList<>();
}
sections.add(cs.getTitle());
userSections.put(userId, sections);
```

bjones86 commented on gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java:
```diff
@@ -2569,6 +2586,15 @@ public String getStudentNumber(final User u, final Site site)
+	public boolean isSectionsVisible() {
+
+		final Optional<Site> site = getCurrentSite();
+		return site.isPresent() && sectionManager.getSections(site.get().getId()).size() > 0;
```
You could do `! .isEmpty()` (not empty) rather than doing a size check here.

bjones86 commented on gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/model/GbUser.java:
```diff
@@ -17,43 +17,46 @@
 
 import java.io.Serializable;
 
+import lombok.experimental.Accessors;
```
This name of this package makes me worry...

adrianfish replied:
I understand, but if lombok remove it, we can implement this another way. Being able to chain ops is a nice feature.

bjones86 commented on gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/component/GbGradeTable.html:
```diff
@@ -238,6 +238,33 @@
                 </div>
             </div>
+        </script>
+				<script id="sectionsHeaderTemplate" type="text/template">
```
Whitespace is a little off here.

adrianfish commented:
About the sakai property: my next task is to make all the gb columns hideable, just like the item columns, so maybe the prop isn't needed.

adrianfish commented:
All done, other than the lombok accessors annotation. I think it's worth taking a chance on that. Travis will catch it if they deprecate it and we update lombok, and it will be easy to code around, imo.

Thanks for the review @bjones86. Some really good catches there!

bjones86 commented:
Yeah, I had the same thought last night after I commented (about the sakai.property). If the 'show/hide column' functionality could be extended to this new column, and perhaps be hidden by default, then it wouldn't bother the people that don't want to see it. Those who do want to see it could toggle it on, and perhaps have this setting be persisted so they don't have to toggle it on every time they visit the gradebook. I think that would be a more feasible solution than the sakai.property.

adrianfish commented:
@bjones86 I've added the property to hide the sections column. Once I've added the column hiding changes we can perhaps revisit taking the prop out. Good to merge?

bjones86 commented:
Sound good, thanks @adrianfish!
